sdk/src/main/scala/com/horizen/account/forger/AccountForgeMessageBuilder.scala

```diff
@@ -110,65 +118,61 @@ class AccountForgeMessageBuilder(
-              return Success(receiptList, listOfTxsInBlock, cumBaseFee, cumForgerTips)
-            case Failure(e: NonceTooLowException) =>
-              // SHOULD NEVER HAPPEN, but in case just skip this tx
-              log.error(s"******** Could not apply tx for NonceTooLowException ******* : ${e.getMessage}")
-              priceAndNonceIter.next()
-            case Failure(e) =>
-              // skip all txs from the same account but remove any changes caused by the rejected tx
-              log.warn(s"Could not forge tx, reason: ${e.getMessage}", e)
-              priceAndNonceIter.removeAndSkipAccount()
-              stateView.revertToSnapshot(revisionId)
-              // Restore gas
-              val usedGas = initialBlockGas.subtract(blockGasPool.getGas)
-              blockGasPool.addGas(usedGas)
-          }
-        } finally {
-          stateView.finalizeChanges()
+        stateView.applyTransaction(tx, listOfTxsInBlock.size, blockGasPool, blockContext) match {
+          case Success(consensusDataReceipt) =>
+            val ethTx = tx.asInstanceOf[EthereumTransaction]
+
+            receiptList += consensusDataReceipt
+            listOfTxsInBlock += tx
+
+            val txGasUsed = consensusDataReceipt.cumulativeGasUsed.subtract(cumGasUsed)
+            // update cumulative gas used so far
+            cumGasUsed = consensusDataReceipt.cumulativeGasUsed
+
+            val (txBaseFeePerGas, txForgerTipPerGas) = GasUtil.getTxFeesPerGas(ethTx, blockContext.baseFee)
+            cumBaseFee = cumBaseFee.add(txBaseFeePerGas.multiply(txGasUsed))
+            cumForgerTips = cumForgerTips.add(txForgerTipPerGas.multiply(txGasUsed))
+            blockSize = blockSize + txSize
+            priceAndNonceIter.next()
+
+          case Failure(err) =>
+            log.trace(s"Could not apply tx, reason: ${err.getMessage}")
+            err match {
+              case _: GasLimitReached =>
+                // block gas limit reached
+                // keep trying to fit transactions into the block: this TX did not fit, but another one might
+                // skip all txs from the same account
+                priceAndNonceIter.removeAndSkipAccount()
+              case _: FeeCapTooLowException =>
+                // stop forging because all the remaining txs cannot be executed for the nonce, if they are from the
+                // same account, or, if they are from other accounts, they will have a lower fee cap
+                return Success(receiptList, listOfTxsInBlock, cumBaseFee, cumForgerTips)
+              case e: NonceTooLowException =>
+                // SHOULD NEVER HAPPEN, but in case just skip this tx
+                log.error(s"******** Could not apply tx for NonceTooLowException ******* : ${e.getMessage}")
+                priceAndNonceIter.next()
+              case e =>
+                // skip all txs from the same account but remove any changes caused by the rejected tx
```

Comment:
"but remove any changes caused by the rejected tx" comment is not actual anymore, because the revert is not anymore processed here, but instead inside `applyTransaction`

Reply:
agree, will remove

Reply:
done